src/Core/Managed/Shared/Extensibility/MetricAggregatorManager.cs

```diff
@@ -0,0 +1,295 @@
+        /// </summary>
+        public MetricAggregatorManager()
+            : this(new TelemetryClient())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MetricAggregatorManager"/> class.
+        /// </summary>
+        /// <param name="client">Telemetry client to use to output aggregated metric data.</param>
+        public MetricAggregatorManager(TelemetryClient client)
+        {
+            this.telemetryClient = client ?? new TelemetryClient();
+            this.aggregatorDictionary = new ConcurrentDictionary<MetricAggregator, SimpleMetricStatisticsAggregator>();
+
+            this.lastSnapshotStartDateTime = DateTimeOffset.UtcNow;
+
+            this.snapshotTask = new TaskTimer() { Delay = GetWaitTime() };
+            this.snapshotTask.Start(this.SnapshotAndReschedule);
+        }
+
+        /// <summary>
+        /// Gets a list of metric processors associated
+        /// with this instance of <see cref="MetricAggregatorManager"/>.
+        /// </summary>
+        internal IList<IMetricProcessor> MetricProcessors
+        {
+            get
+            {
+                TelemetryConfiguration config = this.telemetryClient.TelemetryConfiguration;
+
+                return config.MetricProcessors;
+            }
+        }
+
+        /// <summary>
+        /// Creates single time series metric data aggregator.
+        /// </summary>
+        /// <param name="metricName">Name of the metric.</param>
+        /// <param name="dimensions">Optional dimensions.</param>
+        /// <returns>Value aggregator for the metric specified.</returns>
+        public MetricAggregator CreateMetricAggregator(string metricName, IDictionary<string, string> dimensions = null)
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricAggregatorManager));
+            }
+
+            return new MetricAggregator(this, metricName, dimensions);
+        }
+
+        /// <summary>
+        /// Flushes the in-memory aggregation buffers.
+        /// </summary>
+        public void Flush()
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricAggregatorManager));
+            }
+
+            this.Flush(false);
+        }
+
+        /// <summary>
+        /// Disposes the object.
+        /// </summary>
+        public void Dispose()
+        {
+            this.disposed = true;
+
+            if (this.snapshotTask != null)
+            {
+                this.snapshotTask.Dispose();
+            }
+
+            this.Flush(true);
+        }
+
+        internal SimpleMetricStatisticsAggregator GetStatisticsAggregator(MetricAggregator aggregator)
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricAggregatorManager));
+            }
+
+            if (aggregator == null)
```

Comment:
The `GetStatisticsAggregator` is only called by `MetricAggregator.Track`. Can `aggregator` be ever `null`?